docs/design/services/dynamodb/high-level-library/prototype/option-1/sync/Prototype.java

```diff
@@ -0,0 +1,1026 @@
+         *     <li>{@link InputStream} -> {@link ItemAttributeValueType#BYTES}</li>
+         *     <li>{@link File} -> {@link ItemAttributeValueType#BYTES}</li>
+         *     <li>{@link Boolean} -> {@link ItemAttributeValueType#BOOLEAN}</li>
+         *     <li>{@link Collection} -> {@link ItemAttributeValueType#LIST_OF_*}</li>
+         *     <li>{@link Stream} -> {@link ItemAttributeValueType#LIST_OF_*}</li>
+         *     <li>{@link Iterable} -> {@link ItemAttributeValueType#LIST_OF_*}</li>
+         *     <li>{@link Iterator} -> {@link ItemAttributeValueType#LIST_OF_*}</li>
+         *     <li>{@link Enumeration} -> {@link ItemAttributeValueType#LIST_OF_*}</li>
+         *     <li>{@link Optional} -> {@link ItemAttributeValue#*}</li>
+         *     <li>{@link Map} -> {@link ItemAttributeValueType#ITEM}</li>
+         *     <li>{@link Object} -> {@link ItemAttributeValueType#ITEM}</li>
+         *     <li>{@link null} -> {@link ItemAttributeValueType#NULL}</li>
+         * </ul>
+         *
+         * Usage Example:
+         * <code>
+         *     DocumentClientConfiguration clientConfiguration =
+         *             DocumentClientConfiguration.builder()
+         *                                        .addConverter(InstantsAsStringsConverter.create())
+         *                                        .build();
+         *
+         *     try (DynamoDbDocumentClient client = DynamoDb.documentClientBuilder()
+         *                                                  .documentClientConfiguration(clientConfiguration)
+         *                                                  .build()) {
+         *
+         *         Table table = client.getTable("my-table");
+         *         UUID id = UUID.randomUUID();
+         *         table.putItem(Item.builder()
+         *                           .putAttribute("partition-key", id)
+         *                           .putAttribute("creation-time", Instant.now())
+         *                           .build());
+         *
+         *         Item item = table.getItem(Item.builder()
+         *                                       .putAttribute("partition-key", id)
+         *                                       .build());
+         *
+         *         // Items are usually stored as a number, but it was stored as an ISO-8601 string now because of the
+         *         // InstantsAsStringsConverter.
+         *         assert item.attribute("creation-time").isString();
+         *         assert item.attribute("creation-time").as(Instant.class).isBetween(Instant.now().minus(1, MINUTE),
+         *                                                                            Instant.now());
+         *     }
+         * </code>
+         */
+        DocumentClientConfiguration.Builder converters(List<ItemAttributeValueConverter<?>> converters);
+        DocumentClientConfiguration.Builder addConverter(ItemAttributeValueConverter<?> converter);
```

Comment:
How are converters chosen if more than one is able to convert the item/object?

Reply:
Resolution Path:
1. Iterate over all `contextSatisfies` converters to try to find a match.
2. Iterate over all `isInstanceOf` converts to try to find a match on the class.
3. Iterate over all `isInstanceOf` converters to try to find a match on the parent class.
4. Iterate over all `isInstanceOf` converters to try to find a match on the parent's parent class. (Etc. up to Object which has an SDK-provided converter.)

The result of (2+) can be cached for a given Java type, to prevent having to crawl the chain each time. (1) must occur each time.

When "iterating over" the following precedence applies:
1. Attribute-Level (customer-provided) 
2. Item-Level (customer-provided) 
3. Client-Level (customer provided) 
4. SDK Converters

Reply:
Okay, and for attributes within the same level, whichever is first?

Reply:
Yeah, if they do add(converter1) and then add(converter2), convert1 would be checked first.